src/Confluent.Kafka/Producer.cs

```diff
@@ -671,7 +673,31 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             }
 
             this.ownedKafkaHandle = SafeKafkaHandle.Create(RdKafkaType.Producer, configPtr, this);
-            configHandle.SetHandleAsInvalid(); // config object is no longer usable.
+            configHandle.SetHandleAsInvalid(); // config object ownership was transferred.
+
+            foreach (var partitioner in partitioners)
+            {
+                var topic = partitioner.Key;
+                var topicConfigHandle = SafeTopicConfigHandle.Create();
+                IntPtr topicConfigPtr = topicConfigHandle.DangerousGetHandle();
+                Librdkafka.topic_conf_set_partitioner_cb(topicConfigPtr,
+                    (IntPtr rkt, IntPtr keydata, UIntPtr keylen, int partition_cnt, IntPtr rkt_opaque, IntPtr msg_opaque) =>
+                    {
+                        unsafe
+                        {
+                            var keyIsNull = keydata == IntPtr.Zero;
+                            var keyBytes = keyIsNull
+                                ? ReadOnlySpan<byte>.Empty
+                                : new ReadOnlySpan<byte>(keydata.ToPointer(), (int)keylen);
+                            return partitioners[topic](topic, partition_cnt, keyBytes, keyIsNull);
```

Comment:
Is there any way for the application to get hold of the original Key object in the partitioner?
It shouldn't be available by default (we dont want to hold on to the object due to object lifetime concerns), but some opaque-like construct to pass an arbitrary object to the partitioner (and delivery report delegate) will definitely come in handy.

Reply:
delegates can have associated state, either because they are a closure which captures the state from their context, or because they point to a method on a class instance (which references state in the instance). same with the delivery report handler. there's no need for this to come through librdkafka (which would be more difficult due to lifetimes).

Reply:
But since the partitioner is set on producer instantiation it is not possible to bind any per-message context to the partitioner.

Reply:
What's the use case for that? the user can of course explicitly produce to a partition if more context is required from produce time (which is the work-around currently suggested in lieu of custom partitioners), with the downside being they can't produce before knowing the partition count.

The upgrade path to supporting this if not added now is to add a `PartitionerDelegateWithOpaqe` and overload to the builder + a property to the Message class. Which is all acceptable. I don't think we'll ever need to do it though.